rfcs/text/0003-web-rpc-hooks.md

```diff
@@ -0,0 +1,280 @@
+
+## Pure RPC
+```js
+import {useRPC} from 'fusion-plugin-rpc';
+import {useDispatch, useSelector} from 'react-redux'; 
+
+function Component() {
+  const rpc = useRPC();
+  const dispatch = useDispatch();
+  const user = useSelector(state => state.user.user);
+  if (!user) {
+    throw new Promise((resolve, reject) => {
+      dispatch({type: 'GET_USER_START'});
+      rpc.getUser().then((result) => {
+        dispatch({type: 'GET_USER_SUCCESS', payload: result});
+        resolve();
+      }).catch(e => {
+        dispatch({type: 'GET_USER_FAILURE', payload: e});
+        reject(e);
+      });
+    })
+  }
+  return <div>Hello {user.name}</div>
+}
+```
+
+## RPC + Redux
+```js
+import {useRPCRedux} from 'fusion-plugin-rpc-redux-react';
+import {useDispatch, useSelector} from 'react-redux'; 
+
+function Component() {
+  const rpc = useRPCRedux();
+  const user = useSelector(state => state.user.user);
+  if (!user) {
+    throw rpc.getUser();
+  }
+  return <div>Hello {user.name}</div>
+}
+```
+
+## RPC + Redux + React-Cache Patterns
+```js
+import {createRPCResource} from 'fusion-plugin-rpc-redux-react';
+
+const useUser = createRPCResource('getUser', state => state.user.user);
+
+function Component() {
+  const user = useUser(args);
+  return <div>Hello {user.name}</div>
+}
+```
+
+# Motivation
+
+Hooks and suspense provide opportunities for improved ergonomics of APIs for 
+data fetching and state management. We can take advantage of these best by
+updating our APIs to be compatible with hooks and suspense.
+
+# Detailed design
+
+This RFC proposes a few different APIs at different layers of abstractions. From lowest to highest level of abstraction:
+
+1. `import {useRPC} from 'fusion-plugin-rpc';
+2. `import {useRPCRedux} from 'fusion-plugin-rpc-redux-react';
+3. `import {createRPCResource} from 'fusion-plugin-rpc-redux-react'; (experimental)
+4. `import {createRPCReactorResource} from 'fusion-plugin-rpc-redux-react'; (experimental)
+
+
+## `useRPC`
+
+This API is not designed for direct consumer usage, but rather to be used as a 
+building block into other APIs. 
+
+```js
+import {useRPC} from 'fusion-plugin-rpc';
+
+function Component() {
+  const rpc = useRPC();
+  rpc.getUser(args); // returns a promise
+}
+```
+
+This can be implemented either via a js Proxy or by serializing the supported methods on the server into the html.
+
+## `useRPCRedux`
+
+This can be built using the `useRPC` hook and the `useDispatch` hook. This provides an abstraction that should be easy to use and 
+understand, but does require a small amount of boilerplate. It is designed to be used along side the useSelector API. 
+
+```js
+import {useRPCRedux} from 'fusion-plugin-rpc-redux-react';
+import {useDispatch, useSelector} from 'react-redux'; 
+
+function Component() {
+  const rpc = useRPCRedux();
+  const user = useSelector(state => state.user.user);
+  if (!user) {
+    throw rpc.getUser();
```

Comment:
Will the React team encourage folks to throw promises in their own code? I've only seen examples where thrown promises is purely an implementation detail inside an abstraction.

I'd be wary of encouraging this if it isn't "idiomatic" React.

Reply:
Definitely unclear. I think there needs to be some API that exposes a promise directly. This would be a lower level API that other hooks would build on top of